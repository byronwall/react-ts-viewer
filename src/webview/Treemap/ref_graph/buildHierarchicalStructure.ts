import type { ScopeNode } from "../../../types";
import { findCommonAncestor } from "./graph_nodes";
import { getPathToNode } from "./graph_nodes";

// Helper function to build hierarchical structure preserving parent-child relationships

export function buildHierarchicalStructure(
  nodes: ScopeNode[],
  rootNode: ScopeNode
): ScopeNode {
  const nodeIds = nodes.map((n) => n.id);
  const commonAncestor = findCommonAncestor(rootNode, nodeIds);

  // Build the path from root to the common ancestor so we can include higher-level context
  const ancestorPath = getPathToNode(rootNode, commonAncestor.id);

  // Helper to attach child into shallow copy
  const cloneNode = (n: ScopeNode): ScopeNode => ({
    ...n,
    children: n.children ? [] : [],
  });

  // Build a tree that includes all necessary nodes and their hierarchical relationships
  function buildSubtree(node: ScopeNode): ScopeNode | null {
    // Check if this node should be included
    const shouldInclude = nodeIds.includes(node.id);

    // Process children recursively
    const includedChildren: ScopeNode[] = [];
    if (node.children && node.children.length > 0) {
      for (const child of node.children) {
        const builtChild = buildSubtree(child);
        if (builtChild) {
          includedChildren.push(builtChild);
        }
      }
    }

    // Include this node if either:
    // 1. It's explicitly in our target nodes, OR
    // 2. It has children that need to be included (intermediate node)
    if (shouldInclude || includedChildren.length > 0) {
      const result: ScopeNode = {
        ...node,
        children: includedChildren, // Always use the array, even if empty
      };
      return result;
    }

    return null;
  }

  const hierarchicalTree = buildSubtree(commonAncestor);

  // Now graft this subtree under the ancestor chain to reach the root
  let current: ScopeNode = hierarchicalTree || commonAncestor;
  for (let i = ancestorPath.length - 2; i >= 0; i--) {
    const anc = ancestorPath[i];
    if (!anc) continue;
    const parent = cloneNode(anc);
    parent.children = [current];
    current = parent;
  }

  return current;
}
