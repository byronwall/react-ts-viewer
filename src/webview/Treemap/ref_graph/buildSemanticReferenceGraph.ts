import { type Node, Project, SyntaxKind, ts } from "ts-morph";

import { findInnermostNodeByOffset } from "./graph_nodes";

import type { ScopeNode } from "../../../types";

interface SemanticReference {
  name: string;
  type:
    | "function_call"
    | "variable_reference"
    | "import"
    | "property_access"
    | "destructured_variable";
  sourceNodeId: string;
  targetNodeId?: string;
  position: { line: number; character: number };
  /** Absolute character offset (from SourceFile) where this reference occurs */
  offset: number;
  /** ID of the innermost ScopeNode that contains the reference usage (filled later) */
  usageNodeId?: string;
}

export function buildSemanticReferenceGraph(
  focusNode: ScopeNode,
  rootNode: ScopeNode,
  options: { includeTypeReferences?: boolean } = {}
): {
  nodes: ScopeNode[];
  references: SemanticReference[];
} {
  const { includeTypeReferences = false } = options;

  // Guard against missing or trivial source text
  if (
    !focusNode.source ||
    typeof focusNode.source !== "string" ||
    focusNode.source.trim().length < 3
  ) {
    return {
      nodes: [focusNode],
      references: [],
    };
  }

  /* ------------------------------------------------------------------ *
   * Build ts-morph project for the FULL file so we can resolve        *
   * symbols to their declarations (needed for drawing edges).          *
   * ------------------------------------------------------------------ */

  const fullSource: string =
    typeof rootNode.source === "string" ? rootNode.source : "";

  const project = new Project({ useInMemoryFileSystem: true });

  // Determine a plausible file path from the node id (everything before last ':').
  const match = focusNode.id.match(/^(.*):\d+-\d+$/);
  const sourceFilePath = match ? match[1]! : "file.tsx";

  const sourceFile = project.createSourceFile(sourceFilePath, fullSource, {
    scriptKind: ts.ScriptKind.TSX,
    overwrite: true,
  });

  // Character range of the BOI (focus node) taken directly from the id
  const rangeMatch = focusNode.id.match(/:(\d+)-(\d+)$/);
  const focusStartOffset = rangeMatch ? parseInt(rangeMatch[1]!, 10) : 0;
  const focusEndOffset = rangeMatch
    ? parseInt(rangeMatch[2]!, 10)
    : fullSource.length;

  /* ------------------------------------------------------------------ *
   * Step 1:  Collect identifiers INSIDE focus range + their symbols   *
   * ------------------------------------------------------------------ */

  const declaredInside = new Set<string>();
  const externalUsages: { name: string; node: import("ts-morph").Node }[] = [];

  sourceFile.getDescendantsOfKind(SyntaxKind.Identifier).forEach((id) => {
    const pos = id.getStart();
    if (pos < focusStartOffset || pos > focusEndOffset) return; // outside BOI

    // Check if this identifier *declares* something within BOI
    const parent = id.getParent();
    if (parent) {
      const kind = parent.getKind();
      if (
        (kind === SyntaxKind.VariableDeclaration ||
          kind === SyntaxKind.Parameter ||
          kind === SyntaxKind.FunctionDeclaration ||
          kind === SyntaxKind.ClassDeclaration ||
          kind === SyntaxKind.InterfaceDeclaration ||
          kind === SyntaxKind.TypeAliasDeclaration ||
          kind === SyntaxKind.EnumDeclaration) &&
        id.getStart() === parent.getStart() // being conservative – id is the name
      ) {
        declaredInside.add(id.getText());
        return;
      }
    }

    // ---------------------------------------------
    // Filtering rules:
    // 1. Skip JSX attribute names.
    // 2. Skip intrinsic JSX tag names (opening, self-closing, or closing)
    // 3. Skip identifiers that are the right-hand property in a property-access chain;
    //    we only want the left-most root symbol (e.g., keep "api" in "api.update.mutate").
    // ---------------------------------------------

    const p = id.getParent();
    const pKind = p?.getKind();

    // 1) JSX attribute
    if (pKind === SyntaxKind.JsxAttribute) return;

    // 2) intrinsic JSX tag names (opening, self-closing, or closing)
    if (
      (pKind === SyntaxKind.JsxOpeningElement ||
        pKind === SyntaxKind.JsxSelfClosingElement ||
        pKind === SyntaxKind.JsxClosingElement) &&
      /^[a-z]/.test(id.getText())
    ) {
      return;
    }

    // 3) right-hand side of property access expression
    if (pKind === SyntaxKind.PropertyAccessExpression) {
      // ts-morph provides getNameNode on PropertyAccessExpression
      const pae: any = p;
      if (typeof pae.getNameNode === "function") {
        const nameNode = pae.getNameNode();
        if (nameNode && nameNode.getText() === id.getText()) {
          return; // skip RHS property
        }
      }
    }

    // 4) Identifier used as an explicit property key in an object literal
    //    Example: const obj = { x: 42 } – here `x` should NOT be treated as a
    //    variable reference, only as a literal key. We still want to keep
    //    shorthand properties (e.g., `{ x }`) because in that scenario the
    //    identifier *is* a variable reference.
    if (pKind === SyntaxKind.PropertyAssignment) {
      const pa: any = p;
      if (typeof pa.getNameNode === "function") {
        const nameNode = pa.getNameNode();
        // Skip when the identifier is the *name* of the property assignment
        // (and not part of the initializer).
        if (nameNode && nameNode.getStart() === id.getStart()) {
          return;
        }
      }
    }

    // 5) Skip identifiers that appear only in *type* positions (e.g., in a
    //    type assertion `as SomeType`, generic parameter `<T>`, or other
    //    TypeReference usage) **unless** the caller explicitly opts in via
    //    includeTypeReferences.
    if (!includeTypeReferences) {
      const isTypePosition = (() => {
        let current: Node | undefined = id.getParent();
        const typePositionKinds = new Set([
          SyntaxKind.TypeReference,
          SyntaxKind.TypePredicate,
          SyntaxKind.TypeLiteral,
          SyntaxKind.ImportType,
          SyntaxKind.UnionType,
          SyntaxKind.IntersectionType,
          SyntaxKind.AsExpression,
          SyntaxKind.TypeQuery,
          SyntaxKind.ExpressionWithTypeArguments,
        ]);
        while (current) {
          if (typePositionKinds.has(current.getKind())) return true;
          current = current.getParent();
        }
        return false;
      })();

      if (isTypePosition) {
        return; // skip type-only reference
      }
    }

    // Otherwise, count as external usage
    externalUsages.push({ name: id.getText(), node: id });
  });

  /* ------------------------------------------------------------------ *
   * Step 2:  Convert usages to SemanticReference objects              *
   * ------------------------------------------------------------------ */

  const references: SemanticReference[] = [];
  const nodes: ScopeNode[] = [focusNode];

  const seen = new Set<string>(); // dedupe by name + position

  const offsetToLineChar = (
    src: string,
    off: number
  ): { line: number; character: number } => {
    if (off < 0) return { line: 0, character: 0 };
    const before = src.slice(0, off);
    const line = before.split(/\n/).length; // 1-based
    const lastNewlineIdx = before.lastIndexOf("\n");
    const character = off - (lastNewlineIdx + 1);
    return { line, character };
  };

  for (const { name, node } of externalUsages) {
    if (declaredInside.has(name)) continue; // local capture

    const key = `${name}:${node.getStart()}`;
    if (seen.has(key)) continue;
    seen.add(key);

    let targetNodeId: string | undefined;

    const symbol = node.getSymbol();
    const declarations = symbol ? symbol.getDeclarations() : [];

    if (declarations && declarations.length > 0) {
      const declNode = declarations[0]!;
      const declOffset = declNode.getStart();

      const declScope = findInnermostNodeByOffset(rootNode, declOffset);
      if (declScope && declScope.id !== focusNode.id) {
        targetNodeId = declScope.id;
        if (!nodes.some((n) => n.id === declScope.id)) nodes.push(declScope);
      }
    }

    const offset = node.getStart();
    const position = offsetToLineChar(fullSource, offset);

    // Determine the *innermost* scope node that contains this usage so we can
    // point the arrow to the exact box instead of the BOI header.
    let usageNodeId = focusNode.id;
    const usageScope = findInnermostNodeByOffset(focusNode, offset);
    if (usageScope) {
      usageNodeId = usageScope.id;

      // Ensure this scope is included in the returned node set so the caller
      // can reliably locate it when drawing arrows.
      if (!nodes.some((n) => n.id === usageScope.id)) {
        nodes.push(usageScope);
      }
    }

    references.push({
      name,
      type: "variable_reference",
      sourceNodeId: focusNode.id,
      targetNodeId,
      position,
      offset,
      usageNodeId,
    });
  }

  return {
    nodes,
    references,
  };
}
