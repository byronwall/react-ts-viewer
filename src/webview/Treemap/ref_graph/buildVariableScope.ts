import * as ts from "typescript";
import { VariableScope } from "./graph_nodes";
import { getLineAndCharacter } from "./ts_ast";
import { extractDestructuredNames } from "./ts_ast";

// Build variable scope from AST

export function buildVariableScope(
  node: ts.Node,
  sourceFile: ts.SourceFile,
  parent?: VariableScope
): VariableScope {
  const scope: VariableScope = {
    declarations: new Map(),
    parent,
    level: parent ? parent.level + 1 : 0,
  };

  const walk = (n: ts.Node): void => {
    // Variable declarations (simple identifiers)
    if (
      ts.isVariableDeclaration(n) ||
      ts.isParameter(n) ||
      ts.isFunctionDeclaration(n)
    ) {
      if (n.name && ts.isIdentifier(n.name)) {
        const name = n.name.text;
        const position = getLineAndCharacter(sourceFile, n.name.pos);
        scope.declarations.set(name, { node: n, name, line: position.line });
      }
    }

    // Destructuring patterns â€“ capture each individual name
    if (ts.isVariableDeclaration(n) && n.name) {
      extractDestructuredNames(n.name).forEach((name) => {
        const position = getLineAndCharacter(sourceFile, n.name!.pos);
        scope.declarations.set(name, { node: n, name, line: position.line });
      });
    }

    ts.forEachChild(n, walk);
  };

  walk(node);
  return scope;
}
