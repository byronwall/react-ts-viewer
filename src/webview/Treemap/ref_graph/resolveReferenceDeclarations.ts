import * as ts from "typescript";
import type { ScopeNode } from "../../../types";
import type { SemanticReference } from "./buildSemanticReferenceGraph";
import {
  createSourceFile,
  getLineAndCharacter,
  extractDestructuredNames,
} from "./ts_ast";

/**
 * Information about a declaration corresponding to a {@link SemanticReference}.
 */
export interface ReferenceDeclarationInfo {
  /** Kind of declaration node that introduced the identifier */
  kind:
    | "import"
    | "variable"
    | "destructuring"
    | "parameter"
    | "function"
    | "class"
    | "enum"
    | "interface"
    | "type";
  /** Absolute offset (from beginning of file) where the declaration begins */
  offset: number;
  /** 1-based line/character for human consumption (matches other helpers) */
  position: { line: number; character: number };
}

/**
 * Combines an external {@link SemanticReference} with information about the
 * AST node that *declares* the referenced identifier (if resolvable).
 */
export interface ResolvedReferenceDeclaration {
  reference: SemanticReference;
  declaration: ReferenceDeclarationInfo | null;
}

/**
 * Attempts to resolve each external reference to the *closest* AST node that
 * declares the same identifier within the full source file.  Supported
 * declaration constructs include:
 *   • `import` specifiers (default & named)
 *   • variable declarations (`const foo = …`)
 *   • binding patterns in destructuring (`const { foo } = obj`)
 *   • parameters (`function bar(foo) {}` / `(foo) => {}`)
 *   • function / class / enum / interface / type-alias declarations
 *
 * If a declaration cannot be located, the `declaration` field is `null`.
 * The returned array preserves the original ordering and one-to-one length of
 * the provided `references` array.
 */
export function resolveReferenceDeclarations(
  references: SemanticReference[],
  rootNode: ScopeNode
): ResolvedReferenceDeclaration[] {
  // Bail out early if we have no source code to analyse
  if (!rootNode.source || typeof rootNode.source !== "string") {
    console.warn(
      "⚠️ resolveReferenceDeclarations – rootNode lacks source; unable to resolve declaration sites"
    );
    return references.map((r) => ({ reference: r, declaration: null }));
  }

  const sourceFile = createSourceFile(rootNode.source);

  // Build a map from identifier text → declaration info (first match wins)
  const declarations = new Map<
    string,
    { node: ts.Node; kind: ReferenceDeclarationInfo["kind"] }
  >();

  const addDeclaration = (
    name: string,
    node: ts.Node,
    kind: ReferenceDeclarationInfo["kind"]
  ): void => {
    if (!declarations.has(name)) {
      declarations.set(name, { node, kind });
    }
  };

  /**
   * Walk the AST and capture declaration sites.  We keep the *first* occurrence
   * for each identifier which generally corresponds to the tightest/nearest
   * declaration relative to the top of the file (import → variable, etc.).
   */
  const visit = (node: ts.Node): void => {
    // ---- IMPORT DECLARATIONS ----
    if (ts.isImportDeclaration(node) && node.importClause) {
      const { importClause } = node;
      // Default import → import foo from "…";
      if (importClause.name) {
        addDeclaration(importClause.name.text, importClause.name, "import");
      }
      // Named imports → import { foo, bar as baz } from "…";
      if (
        importClause.namedBindings &&
        ts.isNamedImports(importClause.namedBindings)
      ) {
        importClause.namedBindings.elements.forEach((el) => {
          addDeclaration(el.name.text, el.name, "import");
        });
      }
    }
    // ---- VARIABLE DECLARATIONS (including "const", "let", "var") ----
    else if (ts.isVariableDeclaration(node)) {
      if (ts.isIdentifier(node.name)) {
        addDeclaration(node.name.text, node.name, "variable");
      } else {
        // Handle destructuring patterns { foo } / [foo]
        extractDestructuredNames(node.name).forEach((n) => {
          addDeclaration(n, node.name, "destructuring");
        });
      }
    }
    // ---- PARAMETER DECLARATIONS ----
    else if (ts.isParameter(node)) {
      if (ts.isIdentifier(node.name)) {
        addDeclaration(node.name.text, node.name, "parameter");
      } else {
        extractDestructuredNames(node.name).forEach((n) => {
          addDeclaration(n, node.name, "parameter");
        });
      }
    }
    // ---- FUNCTION DECLARATION ----
    else if (ts.isFunctionDeclaration(node) && node.name) {
      addDeclaration(node.name.text, node.name, "function");
    }
    // ---- CLASS DECLARATION ----
    else if (ts.isClassDeclaration(node) && node.name) {
      addDeclaration(node.name.text, node.name, "class");
    }
    // ---- ENUM DECLARATION ----
    else if (ts.isEnumDeclaration(node) && node.name) {
      addDeclaration(node.name.text, node.name, "enum");
    }
    // ---- INTERFACE DECLARATION ----
    else if (ts.isInterfaceDeclaration(node) && node.name) {
      addDeclaration(node.name.text, node.name, "interface");
    }
    // ---- TYPE ALIAS DECLARATION ----
    else if (ts.isTypeAliasDeclaration(node) && node.name) {
      addDeclaration(node.name.text, node.name, "type");
    }

    ts.forEachChild(node, visit);
  };

  visit(sourceFile);

  // Map each reference to its declaration (if found)
  return references.map((ref) => {
    const decl = declarations.get(ref.name);
    if (!decl) {
      return {
        reference: ref,
        declaration: null,
      };
    }

    const offset = decl.node.getStart(sourceFile);
    const position = getLineAndCharacter(sourceFile, offset);

    return {
      reference: ref,
      declaration: {
        kind: decl.kind,
        offset,
        position,
      },
    };
  });
}
