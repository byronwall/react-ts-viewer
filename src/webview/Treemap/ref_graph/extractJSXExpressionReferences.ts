import * as ts from "typescript";

import { VariableScope } from "./graph_nodes";
import { getLineAndCharacter } from "./ts_ast";
import { isVariableDeclaredInScope } from "./graph_nodes";
import { isIdentifierTypePosition, isKeyword } from "./ts_ast";
import { isIdentifierPartOfJsxAttributeName } from "./ts_ast";
import { SemanticReference } from "./buildSemanticReferenceGraph";

// Helper function to extract references from JSX expressions

export function extractJSXExpressionReferences(
  expression: ts.Expression,
  sourceNodeId: string,
  boiScope: VariableScope,
  references: SemanticReference[],
  sourceFile: ts.SourceFile
) {
  function visitExpression(expr: ts.Expression) {
    const position = getLineAndCharacter(sourceFile, expr.pos);

    if (ts.isIdentifier(expr) && !isIdentifierTypePosition(expr)) {
      // Skip identifiers that are **property names** in:
      //   1. A property access expression – e.g. the `includes` in `obj.includes`.
      //   2. An object-literal property assignment – e.g. the `x` in `{ x: offsetX }`.
      // These are not variable references; we only care about the root identifiers
      // that may need to be passed into the refactored function.
      if (
        (ts.isPropertyAccessExpression(expr.parent) &&
          expr.parent.name === expr) ||
        (ts.isPropertyAssignment(expr.parent) && expr.parent.name === expr)
      ) {
        return; // Ignore property name part
      }

      // Skip identifiers originating from within string literals or template literals.
      if (ts.isStringLiteralLike(expr.parent as ts.Node)) {
        return;
      }

      // Ignore any identifier that belongs to a JSX attribute name (including dashed attributes).
      if (isIdentifierPartOfJsxAttributeName(expr)) {
        return;
      }

      const name = expr.text;
      if (!isKeyword(name)) {
        const isInternal = isVariableDeclaredInScope(name, boiScope);

        if (!isInternal) {
          references.push({
            name,
            type: "variable_reference",
            sourceNodeId,
            position,
            offset: expr.pos,
            isInternal,
            direction: "outgoing",
            targets: [],
          });
        }
      }
    } else if (ts.isPropertyAccessExpression(expr)) {
      // Capture only the root object identifier of the JSX property access
      const rootObj = expr.expression;
      if (ts.isIdentifier(rootObj)) {
        const name = rootObj.text;
        const isInternal = isVariableDeclaredInScope(name, boiScope);
        if (!isInternal) {
          references.push({
            name,
            type: "variable_reference",
            sourceNodeId,
            position,
            offset: expr.pos,
            isInternal,
            direction: "outgoing",
            targets: [],
          });
        }
      }
    } else if (ts.isCallExpression(expr)) {
      if (ts.isIdentifier(expr.expression)) {
        const name = expr.expression.text;
        const isInternal = isVariableDeclaredInScope(name, boiScope);
        if (!isInternal) {
          references.push({
            name,
            type: "function_call",
            sourceNodeId,
            position,
            offset: expr.pos,
            isInternal,
            direction: "outgoing",
            targets: [],
          });
        }
      } else if (ts.isPropertyAccessExpression(expr.expression)) {
        const objectName = ts.isIdentifier(expr.expression.expression)
          ? expr.expression.expression.text
          : "unknown";
        const propertyName = expr.expression.name.text;
        const name = `${objectName}.${propertyName}`;
        const isInternal = isVariableDeclaredInScope(objectName, boiScope);
        if (!isInternal) {
          references.push({
            name,
            type: "property_access",
            sourceNodeId,
            position,
            offset: expr.pos,
            isInternal,
            direction: "outgoing",
            targets: [],
          });
        }
      }
    }

    // Recursively visit child expressions
    ts.forEachChild(expr, (child) => {
      if (ts.isExpression(child)) {
        visitExpression(child);
      }
    });
  }

  visitExpression(expression);
}
