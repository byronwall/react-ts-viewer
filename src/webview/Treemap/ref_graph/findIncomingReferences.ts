import type { ScopeNode } from "../../../types";
import { type SemanticReference } from "./buildSemanticReferenceGraph";
import { type VariableScope } from "./graph_nodes";
import { buildVariableScope } from "./buildVariableScope";
import { extractSemanticReferences } from "./extractSemanticReferences";
import { createSourceFile } from "./ts_ast";

// Find incoming references to the BOI from other nodes

export function findIncomingReferences(
  focusNode: ScopeNode,
  rootNode: ScopeNode,
  boiScope: VariableScope
): SemanticReference[] {
  const incomingRefs: SemanticReference[] = [];
  const boiVariableNames = Array.from(boiScope.declarations.keys());

  if (boiVariableNames.length === 0) {
    return incomingRefs;
  }

  function searchNode(node: ScopeNode) {
    // Skip the focus node itself
    if (node.id === focusNode.id) {
      return;
    }

    if (node.source && typeof node.source === "string") {
      try {
        const sourceFile = createSourceFile(node.source);
        const nodeScope = buildVariableScope(sourceFile, sourceFile);

        // Extract references from this node
        const nodeReferences = extractSemanticReferences(
          sourceFile,
          sourceFile,
          node.id,
          nodeScope
        );

        // Check if any references point to BOI variables
        nodeReferences.forEach((ref) => {
          if (boiVariableNames.includes(ref.name) && !ref.isInternal) {
            incomingRefs.push({
              ...ref,
              direction: "incoming",
              targets: [focusNode.id],
            });
          }
        });
      } catch (error) {
        console.warn(
          `⚠️ Error analyzing node ${node.id} for incoming references:`,
          error
        );
      }
    }

    // Recursively search children
    if (node.children) {
      node.children.forEach(searchNode);
    }
  }

  searchNode(rootNode);
  return incomingRefs;
}
