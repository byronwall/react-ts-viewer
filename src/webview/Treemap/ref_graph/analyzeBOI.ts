import type * as ts from "typescript";
import type { ScopeNode } from "../../../types";
import { type SemanticReference } from "./buildSemanticReferenceGraph";
import { buildVariableScope } from "./buildVariableScope";
import { extractSemanticReferences } from "./extractSemanticReferences";
import { findIncomingReferences } from "./findIncomingReferences";
import { getPathToNode } from "./graph_nodes";
import { createSourceFile, getLineAndCharacter } from "./ts_ast";

interface BOIAnalysis {
  scopeBoundary: { start: number; end: number };
  internalDeclarations: Map<
    string,
    { node: ts.Node; name: string; line: number }
  >;
  externalReferences: SemanticReference[];
  incomingReferences: SemanticReference[];
  recursiveReferences: SemanticReference[];
}

// Analyze Block of Interest (BOI) for semantic references
export function analyzeBOI(
  focusNode: ScopeNode,
  rootNode: ScopeNode
): BOIAnalysis {
  if (!focusNode.source || typeof focusNode.source !== "string") {
    console.warn("⚠️ No source code available for BOI analysis");
    return {
      scopeBoundary: { start: 0, end: 0 },
      internalDeclarations: new Map(),
      externalReferences: [],
      incomingReferences: [],
      recursiveReferences: [],
    };
  }

  try {
    // Create TypeScript source file
    const sourceFile = createSourceFile(focusNode.source);

    // Build variable scope for the BOI
    const boiScope = buildVariableScope(sourceFile, sourceFile);

    // Extract semantic references
    const allReferences = extractSemanticReferences(
      sourceFile,
      sourceFile,
      focusNode.id,
      boiScope
    );

    // ------------------------------------------------------------------
    // Compute absolute position (line/character) across the FULL source file
    // ------------------------------------------------------------------
    const fullSourceFile: ts.SourceFile | null =
      typeof rootNode.source === "string"
        ? createSourceFile(rootNode.source)
        : null;

    // Identify the variable name that *owns* the BOI (e.g. the variable to
    // which an arrow-function is assigned).  Any reference to that variable
    // is considered *internal* and should therefore be excluded from the
    // "external" set.
    const pathToFocus = getPathToNode(rootNode, focusNode.id);
    const owningVarNode = [...pathToFocus].reverse().find((n) => {
      return String(n.category) === "Variable";
    });

    const boiVarName = (() => {
      if (!owningVarNode?.label) return null;
      // Heuristic: first token of the label before whitespace or assignment
      return owningVarNode.label.split(/[\s=:{[(]/)[0] ?? null;
    })();

    // Helper to de-duplicate references (name+type+offset)
    const unique = new Map<string, SemanticReference>();
    allReferences.forEach((ref) => {
      const key = `${ref.name}|${ref.type}|${ref.offset}`;
      if (!unique.has(key)) {
        unique.set(key, ref);
      }
    });

    let dedupedReferences = Array.from(unique.values());

    // Recalculate line/character using the full file so the positions are absolute
    if (fullSourceFile) {
      dedupedReferences = dedupedReferences.map((r) => ({
        ...r,
        position: getLineAndCharacter(fullSourceFile, r.offset),
      }));
    }

    // Categorize references by direction, excluding refs that point to the
    // BOI's own variable name (if detected).
    let externalReferences = dedupedReferences.filter(
      (ref) => !ref.isInternal && ref.name !== boiVarName
    );
    const recursiveReferences = dedupedReferences.filter(
      (ref) => ref.isInternal
    );

    // Final de-duplication for external refs – one entry per (name,type)
    {
      const seen = new Map<string, SemanticReference>();
      externalReferences.forEach((ref) => {
        // not concerned with type here, just want to avoid duplicates
        const key = ref.name;
        if (!seen.has(key)) {
          seen.set(key, ref);
        }
      });
      externalReferences = Array.from(seen.values());
    }

    // Find incoming references by searching the root node for references to BOI variables
    const incomingReferences = findIncomingReferences(
      focusNode,
      rootNode,
      boiScope
    );

    return {
      scopeBoundary: { start: 0, end: focusNode.source.length },
      internalDeclarations: boiScope.declarations,
      externalReferences,
      incomingReferences,
      recursiveReferences,
    };
  } catch (error) {
    console.error("❌ Error in BOI analysis:", error);
    return {
      scopeBoundary: { start: 0, end: 0 },
      internalDeclarations: new Map(),
      externalReferences: [],
      incomingReferences: [],
      recursiveReferences: [],
    };
  }
}
