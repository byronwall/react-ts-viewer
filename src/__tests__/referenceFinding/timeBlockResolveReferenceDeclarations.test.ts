import { describe, expect, it } from "vitest";
import path from "path";

import { analyzeBOI } from "../../webview/Treemap/ref_graph/analyzeBOI";
import { resolveReferenceDeclarations } from "../../webview/Treemap/ref_graph/resolveReferenceDeclarations";
import { createRefGraphObjs } from "./createRefGraphObjs";

describe("reference declaration resolver â€“ TimeBlock.handleMouseDown", () => {
  it("should map each external reference to its declaration", () => {
    const fixturePath = path.join(
      __dirname,
      "..",
      "__fixtures__",
      "TimeBlock.tsx"
    );

    const { focusNode, rootNode } = createRefGraphObjs(
      fixturePath,
      "handleMouseDown"
    );

    const { externalReferences } = analyzeBOI(focusNode, rootNode);

    const resolved = resolveReferenceDeclarations(externalReferences, rootNode);

    // Snapshot to keep output stable and easy to inspect
    expect(resolved).toMatchSnapshot();
  });
});
