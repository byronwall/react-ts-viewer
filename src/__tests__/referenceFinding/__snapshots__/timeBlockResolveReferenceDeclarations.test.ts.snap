// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`reference declaration resolver â€“ TimeBlock.handleMouseDown > should map each external reference to its declaration 1`] = `
[
  {
    "declaration": {
      "kind": "parameter",
      "offset": 912,
      "position": {
        "character": 27,
        "line": 39,
      },
    },
    "reference": {
      "direction": "outgoing",
      "isInternal": false,
      "name": "isPreview",
      "offset": 1218,
      "position": {
        "character": 9,
        "line": 51,
      },
      "sourceNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/TimeBlock.tsx:1165-1717",
      "targets": [],
      "type": "variable_reference",
    },
  },
  {
    "declaration": {
      "kind": "parameter",
      "offset": 912,
      "position": {
        "character": 27,
        "line": 39,
      },
    },
    "reference": {
      "direction": "outgoing",
      "isInternal": false,
      "name": "onResizeStart",
      "offset": 1461,
      "position": {
        "character": 23,
        "line": 60,
      },
      "sourceNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/TimeBlock.tsx:1165-1717",
      "targets": [],
      "type": "function_call",
    },
  },
  {
    "declaration": {
      "kind": "parameter",
      "offset": 912,
      "position": {
        "character": 27,
        "line": 39,
      },
    },
    "reference": {
      "direction": "outgoing",
      "isInternal": false,
      "name": "block",
      "offset": 1482,
      "position": {
        "character": 21,
        "line": 61,
      },
      "sourceNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/TimeBlock.tsx:1165-1717",
      "targets": [],
      "type": "variable_reference",
    },
  },
  {
    "declaration": {
      "kind": "variable",
      "offset": 1092,
      "position": {
        "character": 9,
        "line": 47,
      },
    },
    "reference": {
      "direction": "outgoing",
      "isInternal": false,
      "name": "blockStart",
      "offset": 1498,
      "position": {
        "character": 37,
        "line": 61,
      },
      "sourceNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/TimeBlock.tsx:1165-1717",
      "targets": [],
      "type": "variable_reference",
    },
  },
  {
    "declaration": {
      "kind": "variable",
      "offset": 1130,
      "position": {
        "character": 9,
        "line": 48,
      },
    },
    "reference": {
      "direction": "outgoing",
      "isInternal": false,
      "name": "blockEnd",
      "offset": 1510,
      "position": {
        "character": 49,
        "line": 61,
      },
      "sourceNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/TimeBlock.tsx:1165-1717",
      "targets": [],
      "type": "variable_reference",
    },
  },
  {
    "declaration": {
      "kind": "parameter",
      "offset": 912,
      "position": {
        "character": 27,
        "line": 39,
      },
    },
    "reference": {
      "direction": "outgoing",
      "isInternal": false,
      "name": "onDragStart",
      "offset": 1647,
      "position": {
        "character": 13,
        "line": 64,
      },
      "sourceNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/TimeBlock.tsx:1165-1717",
      "targets": [],
      "type": "function_call",
    },
  },
  {
    "declaration": null,
    "reference": {
      "direction": "outgoing",
      "isInternal": false,
      "name": "x",
      "offset": 1677,
      "position": {
        "character": 30,
        "line": 65,
      },
      "sourceNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/TimeBlock.tsx:1165-1717",
      "targets": [],
      "type": "variable_reference",
    },
  },
  {
    "declaration": null,
    "reference": {
      "direction": "outgoing",
      "isInternal": false,
      "name": "y",
      "offset": 1689,
      "position": {
        "character": 42,
        "line": 65,
      },
      "sourceNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/TimeBlock.tsx:1165-1717",
      "targets": [],
      "type": "variable_reference",
    },
  },
]
`;
